pennylane/devices/qubit/measure.py

```diff
@@ -0,0 +1,132 @@
+"""
+from typing import Callable
+
+from scipy.sparse import csr_matrix
+
+from pennylane import math
+from pennylane.ops import Sum
+from pennylane.measurements import StateMeasurement, MeasurementProcess, ExpectationMP
+from pennylane.typing import TensorLike
+from pennylane.wires import Wires
+
+from .apply_operation import apply_operation
+
+
+def state_diagonalizing_gates(
+    measurementprocess: StateMeasurement, state: TensorLike
+) -> TensorLike:
+    """Apply a measurement to state when the measurement process has an observable with diagonalizing gates.
+
+    Args:
+        measurementprocess (StateMeasurement): measurement to apply to the state
+        state (TensorLike): state to apply the measurement to
+
+    Returns:
+        TensorLike: the result of the measurement
+    """
+    for op in measurementprocess.diagonalizing_gates():
+        state = apply_operation(op, state)
+
+    total_indices = len(state.shape)
+    wires = Wires(range(total_indices))
+    return measurementprocess.process_state(math.flatten(state), wires)
+
+
+def state_hamiltonian_expval(measurementprocess: ExpectationMP, state: TensorLike) -> TensorLike:
+    """Measure the expectation value of the state when the measured observable is a ``Hamiltonian`` or
+    ``SparseHamiltonian``.
+
+    Args:
+        measurementprocess (ExpectationMP): measurement process to apply to the state
+        state (TensorLike): the state to measure
+
+    Returns:
+        TensorLike: the result of the measurement
+    """
+    # Add case for backprop later
+    total_wires = len(state.shape)
+    Hmat = measurementprocess.obs.sparse_matrix(wire_order=list(range(total_wires)))
+    _state = math.toarray(state).flatten()
+
+    # Find the expectation value using the <\psi|H|\psi> matrix contraction
+    bra = csr_matrix(math.conj(_state))
+    ket = csr_matrix(_state[..., None])
+    new_ket = csr_matrix.dot(Hmat, ket)
+    res = csr_matrix.dot(bra, new_ket).toarray()[0]
+    # Add broadcasting later
+
+    return math.real(math.squeeze(res))
+
+
+def get_measurement_function(
+    measurementprocess: MeasurementProcess,
+) -> Callable[[MeasurementProcess, TensorLike], TensorLike]:
+    """Get the appropriate method for performing a measurement.
+
+    Args:
+        measurementprocess (MeasurementProcess): measurement process to apply to the state
+
+    Returns:
+        Callable: function that returns the measurement result
+    """
+    if isinstance(measurementprocess, StateMeasurement):
+        if measurementprocess.obs is None:
+            # no need to apply diagonalizing gates
+
+            def process_state_wrapper(measurementprocess, state):
+                """Wrapper around measurementprocess.process_state to be consistent with
+                function interface."""
+                total_indices = len(state.shape)
+                wires = Wires(range(total_indices))
+                return measurementprocess.process_state(math.flatten(state), wires)
```

Comment:
If we define this outside the `get_measurement` function, we can unit test just `process_state` and we can test that `get_meausrement_function` is dispatching to the correct function.